File: app/src/main/java/foundation/e/blisslauncher/core/customviews/DockGridLayout.java
package foundation.e.blisslauncher.core.customviews;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.drawable.Drawable;
import android.support.annotation.NonNull;
import android.util.AttributeSet;
import android.view.WindowInsets;
import android.widget.GridLayout;

import foundation.e.blisslauncher.BlissLauncher;
import foundation.e.blisslauncher.core.DeviceProfile;
import foundation.e.blisslauncher.core.KotlinUtilsKt;
import foundation.e.blisslauncher.core.blur.BlurWallpaperProvider;
import foundation.e.blisslauncher.core.blur.ShaderBlurDrawable;

public class DockGridLayout extends GridLayout implements Insettable, BlurWallpaperProvider.Listener {

    private Context mContext;

    private final BlurWallpaperProvider blurWallpaperProvider;
    private ShaderBlurDrawable fullBlurDrawable = null;
    private int blurAlpha = 255;

    private final Drawable.Callback blurDrawableCallback = new Drawable.Callback() {
        @Override
        public void invalidateDrawable(@NonNull Drawable who) {
            KotlinUtilsKt.runOnMainThread(() -> {
                invalidate();
                return null;
            });
        }

        @Override
        public void scheduleDrawable(@NonNull Drawable who, @NonNull Runnable what, long when) {

        }

        @Override
        public void unscheduleDrawable(@NonNull Drawable who, @NonNull Runnable what) {

        }
    };

    public DockGridLayout(Context context) {
        this(context, null);
    }

    public DockGridLayout(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public DockGridLayout(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        this.mContext = context;
        setWillNotDraw(false);
        blurWallpaperProvider = BlurWallpaperProvider.Companion.getInstance(context);
        createBlurDrawable();
    }

    @Override
    protected void onAttachedToWindow() {
        super.onAttachedToWindow();

        BlurWallpaperProvider.Companion.getInstance(mContext).addListener(this);
        fullBlurDrawable.startListening();
    }

    @Override
    protected void onDetachedFromWindow() {
        super.onDetachedFromWindow();

        BlurWallpaperProvider.Companion.getInstance(mContext).removeListener(this);
        fullBlurDrawable.stopListening();
    }

    @Override
    protected void onDraw(Canvas canvas) {
        fullBlurDrawable.setAlpha(blurAlpha);
        fullBlurDrawable.draw(canvas);
        super.onDraw(canvas);
    }

    @Override
    protected void onLayout(boolean changed, int left, int top, int right, int bottom) {
        super.onLayout(changed, left, top, right, bottom);
        if (changed) {
            fullBlurDrawable.setBounds(left, top, right, bottom);
        }
    }

    @Override
    public void setInsets(WindowInsets insets) {
        DeviceProfile deviceProfile = BlissLauncher.getApplication(mContext).getDeviceProfile();
        InsettableRelativeLayout.LayoutParams lp = (InsettableRelativeLayout.LayoutParams) getLayoutParams();
        lp.height = deviceProfile.hotseatCellHeightPx + insets.getSystemWindowInsetBottom();
        setPadding(deviceProfile.iconDrawablePaddingPx / 2, 0,
                deviceProfile.iconDrawablePaddingPx / 2, insets.getSystemWindowInsetBottom());
        setLayoutParams(lp);
    }

    private void createBlurDrawable() {
        if (isAttachedToWindow()) {
            fullBlurDrawable.stopListening();
        }
        fullBlurDrawable = blurWallpaperProvider.createDrawable();
        fullBlurDrawable.setCallback(blurDrawableCallback);
        fullBlurDrawable.setBounds(getLeft(), getTop(), getRight(), getBottom());
        if (isAttachedToWindow()) fullBlurDrawable.startListening();
    }

    @Override
    public void onEnabledChanged() {
        createBlurDrawable();
    }

    @Override
    public void onWallpaperChanged() {
    }
}

File: app/src/main/java/foundation/e/blisslauncher/core/customviews/DockGridLayout.kt
package foundation.e.blisslauncher.core.customviews

import android.content.Context
import android.graphics.Canvas
import android.graphics.drawable.Drawable
import android.util.AttributeSet
import android.view.WindowInsets
import android.widget.GridLayout
import foundation.e.blisslauncher.BlissLauncher
import foundation.e.blisslauncher.core.blur.BlurWallpaperProvider
import foundation.e.blisslauncher.core.blur.ShaderBlurDrawable
import foundation.e.blisslauncher.core.runOnMainThread

class DockGridLayout @JvmOverloads constructor(
    private val mContext: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0
) : GridLayout(mContext, attrs, defStyleAttr), Insettable,
    BlurWallpaperProvider.Listener {
    private val blurWallpaperProvider: BlurWallpaperProvider
    private var fullBlurDrawable: ShaderBlurDrawable? = null
    private val blurAlpha = 255
    private val blurDrawableCallback: Drawable.Callback = object : Drawable.Callback {
        override fun invalidateDrawable(who: Drawable) {
            runOnMainThread {
                invalidate()
                null
            }
        }

        override fun scheduleDrawable(
            who: Drawable,
            what: Runnable,
            `when`: Long
        ) {
        }

        override fun unscheduleDrawable(
            who: Drawable,
            what: Runnable
        ) {
        }
    }

    override fun onAttachedToWindow() {
        super.onAttachedToWindow()
        BlurWallpaperProvider.getInstance(mContext).addListener(this)
        fullBlurDrawable!!.startListening()
    }

    override fun onDetachedFromWindow() {
        super.onDetachedFromWindow()
        BlurWallpaperProvider.getInstance(mContext).removeListener(this)
        fullBlurDrawable!!.stopListening()
    }

    override fun onDraw(canvas: Canvas) {
        fullBlurDrawable!!.alpha = blurAlpha
        fullBlurDrawable!!.draw(canvas)
        super.onDraw(canvas)
    }

    override fun onLayout(
        changed: Boolean,
        left: Int,
        top: Int,
        right: Int,
        bottom: Int
    ) {
        super.onLayout(changed, left, top, right, bottom)
        if (changed) {
            fullBlurDrawable!!.setBounds(left, top, right, bottom)
        }
    }

    override fun setInsets(insets: WindowInsets) {
        val deviceProfile = BlissLauncher.getApplication(mContext).deviceProfile
        val lp =
            layoutParams as InsettableRelativeLayout.LayoutParams
        lp.height = deviceProfile.hotseatCellHeightPx + insets.systemWindowInsetBottom
        setPadding(
            deviceProfile.iconDrawablePaddingPx / 2, 0,
            deviceProfile.iconDrawablePaddingPx / 2, insets.systemWindowInsetBottom
        )
        layoutParams = lp
    }

    private fun createBlurDrawable() {
        if (isAttachedToWindow) {
            fullBlurDrawable!!.stopListening()
        }
        fullBlurDrawable = blurWallpaperProvider.createDrawable()
        fullBlurDrawable!!.callback = blurDrawableCallback
        fullBlurDrawable!!.setBounds(left, top, right, bottom)
        if (isAttachedToWindow) fullBlurDrawable!!.startListening()
    }

    override fun onEnabledChanged() {
        createBlurDrawable()
    }

    override fun onWallpaperChanged() {}

    init {
        setWillNotDraw(false)
        blurWallpaperProvider = BlurWallpaperProvider.getInstance(mContext)
        createBlurDrawable()
    }
}